cql3/expr/to_restriction.cc

```diff
@@ -20,14 +20,164 @@
+                                                                           oper_t oper,
+                                                                           expression&& prepared_rhs,
+                                                                           comparison_order order,
+                                                                           const schema& schema) {
+    const column_value& sub_col = get_subscripted_column(lhs_sub);
+    validate_single_column_relation(sub_col, oper, schema, true);
+
+    ::shared_ptr<restrictions::restriction> r = ::make_shared<restrictions::single_column_restriction>(*sub_col.col);
+    r->expression = binary_operator(std::move(lhs_sub), oper, std::move(prepared_rhs));
+    return r;
+}
+
+} // anonymous namespace
+
 ::shared_ptr<restrictions::restriction> to_restriction(const expression& e,
                                                        data_dictionary::database db,
                                                        schema_ptr schema,
                                                        prepare_context& ctx) {
-    throw std::runtime_error(fmt::format("Unimplemented! {}:{}", __FILE__, __LINE__));
+    if (is<constant>(e)) {
+        throw exceptions::invalid_request_exception("Constant value by itself is not supported as a restriction yet");
+    }
+
+    if (!is<binary_operator>(e)) {
+        throw exceptions::invalid_request_exception("Restriction must be a binary operator");
+    }
+    binary_operator binop_to_prepare = as<binary_operator>(e);
+
+    // Convert single element IN relation to an EQ relation
+    if (binop_to_prepare.op == oper_t::IN && is<collection_constructor>(binop_to_prepare.rhs)) {
+        const std::vector<expression>& elements = as<collection_constructor>(binop_to_prepare.rhs).elements;
+        if (elements.size() == 1) {
+            binop_to_prepare.op = oper_t::EQ;
+            binop_to_prepare.rhs = elements[0];
+        }
+    }
+
+    binary_operator prepared_binop = prepare_binary_operator(binop_to_prepare, db, schema, ctx);
+
+    const column_value* lhs_pk_col_search_res = find_in_expression<column_value>(prepared_binop.lhs,
+        [](const column_value& col) {
+            return col.col->is_partition_key();
+        }
+    );
+    auto reset_processing_pk_column = defer([&ctx] () noexcept { ctx.set_processing_pk_restrictions(false); });
+    if (lhs_pk_col_search_res != nullptr) {
+        ctx.set_processing_pk_restrictions(true);
+    }
+    fill_prepare_context(prepared_binop.lhs, ctx);
+    fill_prepare_context(prepared_binop.rhs, ctx);
+
+    if (prepared_binop.op == oper_t::NEQ) {
+        throw exceptions::invalid_request_exception(format("Unsupported \"!=\" relation: {}", prepared_binop));
+    }
+
+    if (prepared_binop.op == oper_t::IS_NOT && !is<null>(prepared_binop.rhs)) {
+        throw exceptions::invalid_request_exception(format("Unsupported \"IS NOT\" relation: {}", prepared_binop));
+    }
```

Comment:
I think we should parse IS NULL and IS NOT NULL into unary operators, not binary operators with a NULL operand. But that can wait.